### DISCUSSION TITLE
Minio multipart upload API did not merge parts in server when completed

### DISCUSSION THREAD

**Discussion by SAILgaosai** [minio/minio#17405] [Q&A] [answered]
According to my understanding, after the files are uploaded in 5MB parts, they should be merged into a complete file storage on the server.

But in fact Minio did not do this. I saw many parts (maybe 100+parts) in the disk used by minio. After calculation, the size of each part is 5MB /12 ≈ 437KB (my minio cluster has 16 hard drives withEC :4)

If Minio merges the files after multipart upload and then stores them, the data blocks on each drive should be file size/12

Because my hard disk is HDD, I think so many parts will increase the inode usage and the number of I/O when GET files. This shortcoming is obvious for HDD. Whether it is a small file or a large file, the small file will enlarge the number of files. , large files will degrade sequential read and write performance.

I don't know if my understanding is correct, thank you for your answer and guidance

**Comment by klauspost** [comment, marked as answer]
Sounds like a problem with your SDK. You don't provide any actual information, so I will not start speculating.

https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html to assemble the object.

https://docs.aws.amazon.com/AmazonS3/latest/API/API_AbortMultipartUpload.html to remove temporary files on failed uploads.

**Reply by SAILgaosai** [reply]
Thanks，I used `mc cp` and `mc mirror` command, I also try Minio Python SDK.
Seems like I need to compose_object [](https://min.io/docs/minio/linux/developers/python/API.html#compose_object)
`compose_object(bucket_name, object_name, sources, )` ?

but I have some confusion about "`my-object-part-one...three`" in the example, It seems that I can directly use the object url to get it after `fput`, so how do I get it to be automatically divided into several parts by the program when I compose him? write directly `my-object.part1`?

```
from minio.commonconfig import ComposeSource
from minio.sse import SseS3

sources = [
    ComposeSource("my-job-bucket", "my-object-part-one"),
    ComposeSource("my-job-bucket", "my-object-part-two"),
    ComposeSource("my-job-bucket", "my-object-part-three"),
]

# Create my-bucket/my-object by combining source object
# list.
result = client.compose_object("my-bucket", "my-object", sources)
```

Also, I didn't find `compose` command in `mc` , what should I do?

**Reply by klauspost** [reply]
It seems like you are overcomplicating things. 

The SDK should handle the multipart upload. Just upload as a single stream and it will split the parts for you.

Uploading as separate objects and copying to a new object is a rather clunky way of uploading.

You can ask on [minio-py](https://github.com/minio/minio-py) project or maybe someone can help you on [slack](https://slack.min.io/).

**Reply by SAILgaosai** [reply]
even I use `mc cp something myminio/bucket`
there will be many part of one file
```
(base) sail@gaosaideMacBook-Air mnt % mc tree --files disk1/bucket/something/
disk1/bucket/something/
├─ 26eef643-c639-4d9f-8d24-7a5ff7e77245
│  └─ part.1
├─ 494eb880-6aab-4d00-a0b6-58cf65d7128f
│  ├─ part.1
│  ├─ part.10
│  ├─ part.100
│  ├─ part.101
│  ├─ part.102
│  ├─ part.103
│  ├─ part.104
│  ├─ part.105
│  ├─ part.106
│  ├─ part.107
│  ├─ part.108
│  ├─ part.109
...
│  ├─ part.92
│  ├─ part.93
│  ├─ part.94
│  ├─ part.95
│  ├─ part.96
│  ├─ part.97
│  ├─ part.98
│  └─ part.99
├─ 7cdd4c52-876b-4d50-af16-a1ac48487341
│  └─ part.1
└─ xl.meta
```

`7cdd4c52-876b-4d50-af16-a1ac48487341->part.1` was uploaded by`mc cp --disable-multipart something myminio/bucket`
`26eef643-c639-4d9f-8d24-7a5ff7e77245->part.1` was uploaded by web minio-console
one part per drive per file, this is what I expected.
But in my opinion,`--disable-multipart` also not a friendly solution, Multipart uploads should be done automatically and merged on the server.

**Reply by klauspost** [reply]
You should never touch backend files. They are not readable and not meant to be.

**Reply by SAILgaosai** [reply]
I stat the backend files on ubuntu server with minio `Version:2023-06-09T07:32:12Z` mount one xfs 8T HDD drive

```
$ xfs_info /dev/sdc
meta-data=/dev/sdc               isize=512    agcount=32, agsize=61042688 blks
         =                       sectsz=512   attr=2, projid32bit=1
         =                       crc=1        finobt=1, sparse=1, rmapbt=0
         =                       reflink=1
data     =                       bsize=4096   blocks=1953366016, imaxpct=5
         =                       sunit=64     swidth=64 blks
naming   =version 2              bsize=4096   ascii-ci=0, ftype=1
log      =internal log           bsize=4096   blocks=521728, version=2
         =                       sectsz=512   sunit=64 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
$ stat part.1 
  File: part.1
  Size: 16777728        Blocks: 32776      IO Block: 4096   regular file
Device: 820h/2080d      Inode: 11811419394  Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2023-06-10 21:49:01.338496309 +0000
Modify: 2023-06-10 21:49:02.850490565 +0000
Change: 2023-06-10 21:49:02.850490565 +0000
 Birth: -
$ stat part.2
  File: part.2
  Size: 16777728        Blocks: 32776      IO Block: 4096   regular file
Device: 820h/2080d      Inode: 12885092812  Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2023-06-10 21:49:04.366484804 +0000
Modify: 2023-06-10 21:49:05.906478948 +0000
Change: 2023-06-10 21:49:05.906478948 +0000
 Birth: -
$ stat part.3
  File: part.3
  Size: 16777728        Blocks: 32776      IO Block: 4096   regular file
Device: 820h/2080d      Inode: 13421958594  Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2023-06-10 21:49:08.098470611 +0000
Modify: 2023-06-10 21:49:10.078463075 +0000
Change: 2023-06-10 21:49:10.082463060 +0000
 Birth: -

$ sudo xfs_db -r /dev/sdc
xfs_db> inode 11811419394
xfs_db> bmap
data offset 0 startblock 1476533376 (22/138368) count 4097 flag 0
xfs_db> inode 12885092812
xfs_db> bmap
data offset 0 startblock 1610747008 (24/134272) count 4097 flag 0
xfs_db> inode 13421958594
xfs_db> bmap
data offset 0 startblock 1677867840 (25/146240) count 4097 flag 0
```

The conclusion is that the physical addresses of these parts are not continuous. Although they are allocated to different AGs (Allocation Groups).
So is it **faster** to read AGs in parallel than to read a continuous address space when getting an object?

I care about these details because my minio cluster mainly uses HDD instead of SSD
